node/bft/src/helpers/cache_round.rs

```diff
@@ -0,0 +1,248 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use anyhow::{ensure, Result};
+use snarkvm::{
+    console::types::{Address, Field},
+    ledger::committee::Committee,
+    prelude::Network,
+};
+use std::collections::HashSet;
+
+#[derive(Copy, Clone, Debug)]
+struct AddressWithCoordinate<N: Network> {
+    address: Address<N>,
+    x: Field<N>,
+}
+
+impl<N: Network> From<Address<N>> for AddressWithCoordinate<N> {
+    fn from(address: Address<N>) -> Self {
+        Self { address, x: address.to_group().to_x_coordinate() }
+    }
+}
+
+#[derive(Debug)]
+pub struct RoundCache<N: Network> {
+    /// The current highest round which has (stake-weighted) quorum
+    last_highest_round_with_quorum: u64,
+    /// A list of (round, Vec<AddressWithCoordinate<N>>), indicating the last seen highest round for each address
+    highest_rounds: Vec<(u64, Vec<AddressWithCoordinate<N>>)>,
+    /// A list of (AddressWithCoordinate<N>, round) to quickly find an Address' round by their x coordinate
+    address_rounds: Vec<(AddressWithCoordinate<N>, u64)>,
+}
+
+impl<N: Network> Default for RoundCache<N> {
+    /// Initializes a new instance of the cache.
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<N: Network> RoundCache<N> {
+    /// Initializes a new instance of the cache.
+    pub fn new() -> Self {
+        Self {
+            last_highest_round_with_quorum: Default::default(),
+            highest_rounds: Default::default(),
+            address_rounds: Default::default(),
+        }
+    }
+
+    /// Insert a validator at a round
+    fn insert_validator_at_round(&mut self, round: u64, validator: AddressWithCoordinate<N>) {
```

Comment:
Why doesn't this add the validator to `address_rounds`?

Reply:
The function is badly named, we are only updating because we saw a new round for this validator: ef446e9f2